zp-relayer/services/fee/OptimismFeeManager.ts

```diff
@@ -0,0 +1,81 @@
+import type Web3 from 'web3'
+import type BN from 'bn.js'
+import type { Contract } from 'web3-eth-contract'
+import { OP_GAS_ORACLE_ADDRESS, MOCK_CALLDATA } from '@/utils/constants'
+import { AbiItem, toBN } from 'web3-utils'
+import OracleAbi from '@/abi/op-oracle.json'
+import { contractCallRetry } from '@/utils/helpers'
+import {
+  FeeManager,
+  DefaultFeeEstimate,
+  IFeeEstimateParams,
+  IFeeManagerConfig,
+  IUserFeeOptions,
+  IGetFeesParams,
+} from './FeeManager'
+import { IPriceFeed } from '../price-feed'
+
+const ONE_BYTE_GAS = 16
+
+class OptimismUserFeeOptions implements IUserFeeOptions {
+  constructor(private fee: BN, private oneByteFee: BN) {}
+
+  applyFactor(factor: BN) {
+    this.fee = this.fee.mul(factor).divn(100)
+    this.oneByteFee = this.oneByteFee.mul(factor).divn(100)
+    return this
+  }
+
+  denominate(denominator: BN): this {
+    this.fee = this.fee.div(denominator)
+    this.oneByteFee = this.oneByteFee.div(denominator)
+    return this
+  }
+
+  async convert(priceFeed: IPriceFeed) {
+    const [fee, oneByteFee] = await priceFeed.convert([this.fee, this.oneByteFee])
+    this.fee = fee
+    this.oneByteFee = oneByteFee
+    return this
+  }
+
+  getObject() {
+    return {
+      fee: this.fee.toString(10),
+      oneByteFee: this.oneByteFee.toString(10),
+    }
+  }
+}
+
+export class OptimismFeeManager extends FeeManager {
+  private oracle: Contract
+
+  constructor(config: IFeeManagerConfig, web3: Web3) {
+    super(config)
+    this.oracle = new web3.eth.Contract(OracleAbi as AbiItem[], OP_GAS_ORACLE_ADDRESS)
+  }
+
+  async _estimateFee({ gasLimit, data }: IFeeEstimateParams) {
+    const l2Fee = await this.estimateExecutionFee(gasLimit)
+    const l1Fee = await contractCallRetry(this.oracle, 'getL1Fee', [data]).then(toBN)
```

Comment:
should `data` here be passed together with other transaction's data or their mocks (except, the transaction signature)?

Reply:
Yes, will update it soon

Reply:
I added a constant to account for RLP encoded tx overhead. While it may be not 100% accurate and may vary depending on `chainId` or `gasPrice` length, these fluctuations are not significant.